### PR TITLE
[orc8r][helm] Refactor orc8r chart with cert-manager

### PR DESCRIPTION
**Pull request by ShubhamTatvamasi**
## Summary

For Issue: https://github.com/magma/magma/issues/5383
With `certs` helm chart these certificates are auto-generated by cert-manager.

- admin-operator-tls
- certifier-tls
- controller-tls
- bootstrapper-tls
- fluentd-tls
- root-tls
- nms-tls

These new secrets are currently **not** used by other charts.

## Test Plan

Install prerequisite:
```bash
helm repo add jetstack https://charts.jetstack.io
helm repo update

helm install cert-manager jetstack/cert-manager \
  --create-namespace \
  --namespace cert-manager \
  --set installCRDs=true
```

Install certs:
```bash
cd $MAGMA_ROOT/orc8r/cloud/helm/orc8r/charts/certs
helm install certs .
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/values.yaml]: For here and other values.yml file: current default is 1 year, I think sticking to that default is sensible

File: CODEOWNERS
Comment (hcgatewood) [inline CODEOWNERS]: This change would need to be in a separate PR

Updating the codeowners is equivalent to updating the set of project maintainers. See https://www.magmacore.org/community/ for context on that process

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/values.yaml:27]: Here and for other values.yml: I think it makes sense to default custom issue to let's encrypt -- are there any downsides to this approach?

File: orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-pkcs12-password.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-pkcs12-password.yaml]: I think it makes sense to make this var required. Easiest way to make sure everyone uses "password" as their password is to make it the default

File: orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml]: These hard-coded 10-year values should probably be variables. Related: what is the max duration supported for this value?

This is probably also a good point to consider what happens after they expire. The problem is we provision gateways with the specific root and fluentd crts, so when they expire under our current architecture, shit goes down. Specifically

- Root
    - All gateways drop connection to Orc8r, requiring reprovisioning of all gateways in network
        - Solution 1: support in-band mechanism for Orc8r sending new rootCA.pem to all gateways (messy/what if a gateway was disconnected from Orc8r for long period of time)
        - Solution 2: remove support for self-signed certs -- this way, gateways wouldn't need to have a rootCA.pem, instead relying on OS-level support
- Fluentd
    - Same story as root but just for log aggregation. Probably same solution too
- Certifier
    - Should be fine if the underlying keypair remains the same

**Review by hcgatewood — COMMENTED** [review]
Excited for this. Couple general notes

- Is the general plan to create these secrets with new names, then gradually transition application code to consuming these cloud-native versions of the secrets?
- We should be able to rotate every secret except for
    - rootCA key+crt, since the rootCA.pem is manually provisioned onto each gateway... though see my note for how we could get around this
    - certifier key+crt, since that would effectively flush all admin_operator crts... though perhaps doing this infrequently but regularly would actually be a good idea? cc @karthiksubraveti @mpgermano 
- However: when a secret is rotated, we'd need to bounce the services that consume (the old version of) that secret. I think there should be an easy way to automate that, e.g. https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments

File: CODEOWNERS
Comment (ShubhamTatvamasi) [reply CODEOWNERS]: Okay, I will remove this for now.

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/values.yaml:27]: By default, most users use letsencrypt with **HTTP-01 challenge** which is okay for a single public domain.

But we have a wild card sub-domain for NMS which requires **DNS-01 challenge** for gathering certificate which requires updating DNS entries by cert-manager. And not all Domain providers are supported by cert-manager.

That's why I think we should leave it to the user for this setup.

File: orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-pkcs12-password.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-pkcs12-password.yaml]: Okay, Will update this

File: orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml]: We don't have to worry about the expiration of these certificates. cert-manager will auto-renew those.
Even if the `secret` is deleted it will just create a new secret with a fresh certificate.

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/values.yaml]: It's 10 years

https://github.com/magma/magma/blob/master/orc8r/cloud/deploy/scripts/self_sign_certs.sh#L42

**Comment by ShubhamTatvamasi** [comment]
@hcgatewood if we just set the low `duration`  time it will automatically renew these certificates.

We can also set `renewBefore: 12h` or any custom time to rotate it before expiration.

File: orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml]: I don't know the max duration, but I was able to generate 100 years duration certificate.

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (hcgatewood) [reply orc8r/cloud/helm/orc8r/values.yaml]: Oh wow I 100% thought it was 1 year. Good to know

File: orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml
Comment (hcgatewood) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml]: So my concern isn't on the Orc8r side -- I agree that everything should work fine, from a cert perspective, on Orc8r side. The issue actually is we manually copy the root cert to every AGW that will connect to the Orc8r.

This means that once the *original* root cert expires, the gateways will immediately view Orc8r controller certs as invalid. And even though Orc8r has renewed its root cert, we don't currently have a mechanism to propagate that renewed cert down to each gateway

Let me know if I'm missing something

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (hcgatewood) [reply orc8r/cloud/helm/orc8r/values.yaml]: I'm fine with leaving default relatively high, but since auto-renew should work well, something like 1 or 3 years for these non-root certs also seems reasonable. Those original values were chosen under the expectation that it would be a hassle to manually renew

That aside, we do want to rotate the bootstrapper.key frequently, e.g. on a monthly or yearly basis, since it's being used relatively frequently (once every 4 days x number of gateways pointing to the Orc8r)

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (hcgatewood) [reply orc8r/cloud/helm/orc8r/charts/certs/values.yaml:27]: I see, thanks for providing context. I think that's a sensible decision then

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (karthiksubraveti) [inline orc8r/cloud/helm/orc8r/charts/certs/values.yaml]: minor nit: Let's just call this domain_name, calling it dns domain sounds redundant.

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (karthiksubraveti) [inline orc8r/cloud/helm/orc8r/values.yaml:55]: @hcgatewood This probably obsoletes the secret section above right ?

**Review by karthiksubraveti — COMMENTED** [review]
@hcgatewood @ShubhamTatvamasi I'm still trying to understand how cert-manager works here. Can you help me understand what's the operator's flow here. When are the certs generated and how are they pushed to the individual pods ? 

Also we have a orc8r-secrets chart, Does that get obsoleted by this addition ?

@hcgatewood When I was reading through cert manager docs, It seems like our current gateway bootstrapping mechanism also can be replaced by this.  Is my understanding right ?

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (arunuke) [reply orc8r/cloud/helm/orc8r/values.yaml]: Agree with @hcgatewood on the need for rotation. The CA/Browser forum officially [adopted ](https://thehackernews.com/2020/09/ssl-tls-certificate-validity-398.html)398 days as the time limit for expiring security certificates. We should aspire to meet this goal with the option to automatically check and renew security certificates, which cannot happen with self-signed certificates since they do not share ancestry with any other certificates

File: orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml
Comment (arunuke) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/certifier-certificate.yaml]: The possible resolution for this requires that only public CA signed certificates be used over self-signed (which cannot be automatically renewed). The classic flow would be to (a) Using an API, request for a public CA certificate that is used as the root to generate all the other keys. AWS Certificate Manager provides public CA certificates for free. (b) In the absence of an API based request or certificates already available, use public CA certificates from a local directory to generate all the other certificates (c) set expiry to 398 days for certificates (d) use cert manager to renew before expiry.

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/values.yaml]: I have added a new `bootstrapper-certificate`, which will create a new RSA key for us and also added `rotationPolicy: Always` to it.

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/values.yaml]: Updated the name to `domainName`

File: orc8r/cloud/helm/orc8r/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/values.yaml:55]: No, we are not using any of these auto-generated secrets inside pods.

**Comment by ShubhamTatvamasi** [comment]
@mattymo can you please also have a look at this PR?

**Review by mattymo — APPROVED** [review]

**Comment by hcgatewood** [comment]
Mirroring my comments from Slack: I'm not ready to merge this PR until we have a clear story for how we're handling cloud-native Orc8r certs

This PR introduces useful new affordances for how Orc8r handles its certificates, but I think we'll miss out on concrete impact if we don't start with a clear high-level story for cloud-native Orc8r certs -- specifically since these changes will affect AGWs as well.

I'll write up a short GH Discussion proposal this week and send out a cal invite for us to align on how to proceed

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Clearing my review queue. See https://github.com/magma/magma/discussions/5728#discussioncomment-556134 for comments on next steps in getting this PR merged

**Comment by ShubhamTatvamasi** [comment]
@hcgatewood I have added the following things:
1. let's encrypt issuers
1. rotation timings per secret

But the default is not set to let's encrypt issuer, because **HTTP-01 challenge** doesn't support wilcard certificates.

File: orc8r/cloud/helm/orc8r/charts/certs/templates/bootstrapper-certificate.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/templates/bootstrapper-certificate.yaml:14]: Bootstrapper is just a private key -- no certificate necessary

(That said, if it's easier to have it be a cert instead of just a private key, that's fine. Just add a comment indicating as such.)

File: orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-certificate.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-certificate.yaml:16]: Minor nit: perhaps this is the right time to start calling these admin operator certs `default_admin_operator`. We support any number of admin operator certificates (i.e. certs giving admin permissions to the rest api), this is just the default one we create during provisioning

Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-certificate.yaml:16]: Side note for a future PR: we'll still need to inject this default cert into the certifier service https://magma.github.io/magma/docs/next/orc8r/deploy_install#create-an-orchestrator-admin-user

Though perhaps we can invert this requirement -- instead of it being a manual step during installation, the certifier service can instead proactively check for the default-admin-operator-tls secret and pull it in if it exists

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/values.yaml:46]: Let's also add the note here about using the provided letsencrypt issuers

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Looking good, I think we're almost there. Couple top-level things

### ACME challenge type

I don't understand the reasoning for supporting HTTP01 -- my cursory understanding is that it won't work at all for NMS (?).

To me, the goal here is to provide a *default that just works*. Supporting DNS01 via Let's Encrypt allows us just that -- a default that "just works". Let me know if I'm misunderstanding something.

So then we have the following deployment types

- (a) ACME (managed) (default)
	- Default via Let's Encrypt
	- Also support for non-default providers via Helm values
- (b) Self-signed (managed)
- (c) Traditional (unmanaged)
	- This is the current mechanism
	- Upload arbitrary secrets via AWS KMS, manually inject into the cluster via Terraform

I don't think (a) works unless we provide DNS01 challenge type for our Let's Encrypt provider. Looks like (b) works great given what you've written, and of course (c) works bc it's our existing functionality.

### Configuring deployment type

I think keeping these helm charts configurable, as you have them now, is the right choice. Then, when we update Terraform with the option to deploy cert-manager, we can provide better guidance on selecting between the 3 above-mentioned deployment types.

**Comment by hcgatewood** [comment]
From offline sync

- For now, we will default to self-signed issuers and forego support for ACME-based issuers
	- HTTP01: NMS requires wildcard certs, and we don't want to create the footgun of requiring a kubectl command in order to create a new NMS tenant
	- DNS01: Orc8r doesn't currently manage its DNS records. @ShubhamTatvamasi will take on writing up a proposal for remedying this, deprioritized till after we integrate this existing PR with Orc8r/NMS as a whole
- This restricts us to the following deployment types: managed self-signed and unmanaged traditional. This will still be a win, however, because (a) for self-signed deployments, cert management will be more straightforward, and (b) this PR paves the way for, once we wrangle the DNS management story, having fully-managed+fully-trusted certs generated via DNS01 method

File: orc8r/cloud/helm/orc8r/charts/certs/values.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/values.yaml:46]: Removed letsencrypt issuers as per our discussion.

File: orc8r/cloud/helm/orc8r/charts/certs/templates/bootstrapper-certificate.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/bootstrapper-certificate.yaml:14]: I have updated the same on `values.yaml` file.

File: orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-certificate.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-certificate.yaml:16]: For now let it be as it is, so we don't get confused and also it's compatible with previous versions.

Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/admin-operator-certificate.yaml:16]: I will look into it next once this PR has merged.

File: orc8r/cloud/helm/orc8r/charts/certs/templates/controller-certificate.yaml
Comment (hcgatewood) [inline orc8r/cloud/helm/orc8r/charts/certs/templates/controller-certificate.yaml:33]: Why are we having controller.crt and nms.crt each cover both controller and nms domains?

**Review by hcgatewood — APPROVED** [review]
Lgtm. One small q then let's merge

File: orc8r/cloud/helm/orc8r/charts/certs/templates/controller-certificate.yaml
Comment (ShubhamTatvamasi) [reply orc8r/cloud/helm/orc8r/charts/certs/templates/controller-certificate.yaml:33]: Keep both for backwards compatibilities.

**Pull request merged**
